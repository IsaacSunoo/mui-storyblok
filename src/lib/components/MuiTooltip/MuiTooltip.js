import React, { createElement } from 'react';
import PropTypes from 'prop-types';
import Tooltip from '@material-ui/core/Tooltip';
import IconButton from '@material-ui/core/IconButton';

import Storyblok from '../../utils/Storyblok';
import MuiIconButtonRedirect from '../MuiIconButtonRedirect/MuiIconButtonRedirect';
import MuiIconButtonHref from '../MuiIconButtonHref/MuiIconButtonHref';
import MuiIconButtonDownload from '../MuiIconButtonDownload/MuiIconButtonDownload';
import MuiIconButtonDialog from '../MuiIconButtonDialog/MuiIconButtonDialog';
import MuiButtonRedirect from '../MuiButtonRedirect/MuiButtonRedirect';
import MuiButtonHref from '../MuiButtonHref/MuiButtonHref';
import MuiButtonDownload from '../MuiButtonDownload/MuiButtonDownload';
import MuiButtonDialog from '../MuiButtonDialog/MuiButtonDialog';

const MuiTooltip = ({
  rootClass,
  title,
  arrow,
  placement,
  content,
}) => {
  const components = {
    MuiIconButtonRedirect,
    MuiIconButtonHref,
    MuiIconButtonDownload,
    MuiIconButtonDialog,
    MuiButtonRedirect,
    MuiButtonHref,
    MuiButtonDownload,
    MuiButtonDialog,
  };

  const styles = Storyblok.arrayToMuiStyles(rootClass);
  return (
    <Tooltip
      className={styles.root}
      placement={placement}
      title={title}
      arrow={arrow}
    >
      <IconButton>
        {/*
          Warning: validateDOMNesting(...): <button> cannot appear as a descendant of <button>
          components use IconButton to wrap MuiIcon and handle onClick event but when its
          rendered Tooltip is not able to show message
        */}
        {(createElement(components[content[0].component], content[0]))}
      </IconButton>
    </Tooltip>
  );
};

export default MuiTooltip;

MuiTooltip.propTypes = {
  /** stroyblok multiselect of css classes */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /** text for tooltip */
  title: PropTypes.string.isRequired,
  /**
   * mui prop: true | false
   * If true, adds an arrow to the tooltip.
   * */
  arrow: PropTypes.bool,
  /**
   * mui prop: 'bottom-end' | 'bottom-start' | 'bottom' | 'left-end' | 'left-start' | 'left' | 'right-end' | 'right-start' | 'right' | 'top-end' | 'top-start' | 'top'
   * Tooltip placement.
   * */
  placement: PropTypes.string,

  /** Allowed maximum: 1
   *  MuiIconButtonRedirect,
    MuiIconButtonHref,
    MuiIconButtonDownload,
    MuiIconButtonDialog,
    MuiButtonRedirect,
    MuiButtonHref,
    MuiButtonDownload,
    MuiButtonDialog,
  */
  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,
};

MuiTooltip.defaultProps = {
  arrow: false,
  placement: 'bottom',
  rootClass: [],
};
