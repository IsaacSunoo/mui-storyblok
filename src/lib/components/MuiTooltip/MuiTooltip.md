MuiTooltip example:

```js noeditor
import MuiTooltip from './MuiTooltip';
```

```js
const icon = {
    component: 'MuiIcon',
    iconName: 'done_all',
};

const iconButtonDownload = {
    component: 'MuiIconButtonDownload',
    herf: "https://miro.medium.com/max/1200/1*mk1-6aYaf_Bes1E3Imhc0A.jpeg",
    fileName: "cool.jpeg",
    icon: [icon],
};

const content = [iconButtonDownload];

<MuiTooltip
  title="tool tip title"
  content={content}
/>
```