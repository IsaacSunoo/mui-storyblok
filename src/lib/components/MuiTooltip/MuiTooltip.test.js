import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import MuiTooltip from './MuiTooltip';

function setup() {
  const props = {
    title: 'title text',
    content: [{
      component: 'MuiIconButtonHref',
      href: 'wasd.com',
      iconButton: [{
        component: 'MuiIconButton',
        icon: [{
          component: 'MuiIcon',
          iconName: 'star',
        }],
      }],
    }],
  };
  const comp = shallow(<MuiTooltip {...props} />);
  return { comp, props };
}

describe('<MuiTooltip />', () => {
  it('renders MuiTooltip', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();

    const tree = renderer.create((
      <MemoryRouter>
        <MuiTooltip {...props} />
      </MemoryRouter>
    ));
    expect(tree).toMatchSnapshot();
  });
});
