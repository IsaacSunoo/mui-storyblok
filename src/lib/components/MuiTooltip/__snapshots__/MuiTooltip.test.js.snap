// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiTooltip /> snapshot 1`] = `
<button
  aria-describedby={null}
  className="MuiButtonBase-root MuiIconButton-root makeStyles-root-3"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  title="android"
  type="button"
>
  <span
    className="MuiIconButton-label"
  >
    <div />
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;
